open-api/rest-catalog-open-api.yaml

```diff
@@ -3103,6 +3103,81 @@ components:
         uuid:
           type: string
 
+    ADLSCredentials:
+      type: object
+      allOf:
+        - $ref: '#/components/schemas/Credentials'
+      required:
+        - type
+      properties:
+        type:
+          type: string
+          enum: [ "adls" ]
+        account-name:
+          type: string
+        account-key:
+          type: string
+        sas-token:
+          type: string
+        expires-at-ms:
+          type: integer
+          format: int64
+
+    GCSCredentials:
+      type: object
+      allOf:
+        - $ref: '#/components/schemas/Credentials'
+      required:
+        - type
+        - token
+        - expires-at-ms
+      properties:
+        type:
+          type: string
+          enum: [ "gcs" ]
+        token:
+          type: string
+        expires-at-ms:
+          type: integer
+          format: int64
+
+    S3Credentials:
+      type: object
+      allOf:
+        - $ref: '#/components/schemas/Credentials'
+      required:
```

Comment:
Could we add the `region`? Without it, clients might run into issues. Here’s why it would be helpful:
1. Setting a default region on the client side doesn’t work well since clients could be accessing s3 files from different regions.
2. Clients could try to figure out the region using `get-bucket-location`, but this means an extra API call. Plus, if the  vended credentials don’t allow this call, it could cause more problems.

Adding the `region` would simplify things and enhance overall client reliability.

Reply:
+1 for including `region`.

Reply:
Despite the [other comment below](https://github.com/apache/iceberg/pull/10722/files#r1755423596) - the region should be the same as the one for the S3FileIO configuration?

Reply:
@flyrain the region is already sent back in the `config` itself, so I'm not sure why we would want to make it part of `S3Credentials? The region is for configuring the S3 client but the credentials are for configuring `S3FileIO`, so it's a different layer where the region is being used

Reply:
I don't think the region is bound to the credential.  I believe it's possible to build a policy that allows access to bucket in multiple regions, so region should be associated with the bucket, not the credentials.

Reply:
> the region is already sent back in the config of LoadTableResult/ LoadViewResult itself

> I don't think the region is bound to the credential. I believe it's possible to build a policy that allows access to buckets in multiple regions, so region should be associated with the bucket, not the credentials. Therefore, we shouldn't include it here.

I agree that the region is not bound to the credentials. I thought we had moved everything out of the config, didn’t we? I’m okay with placing it in different locations, but currently, the region in config lacks a defined schema. For consistency, perhaps we should add a schema for it.

Reply:
> I thought we are moving everything out of the config, didn’t we? 

@flyrain we're only standardizing credentials and not everything that we currently pass via the `config`. Standardizing all configuration is a bigger task and warrants a separate proposal. Besides that it's worth mentioning that a server would still have to sent credentials via the `config` for older clients.

Reply:
Thanks for clarification, @nastra ! 
Does it make sense to introduce a `storage` property within LoadTableResult? By doing that, you can group all storage-related settings—such as type, config (e.g., region), and credentials—into a cohesive structure. This approach is especially beneficial if we plan to support multiple storage types or need to extend configurations in the future.

Here's how the updated schema might look:
```
LoadTableResult:
  properties:
    storage:
      $ref: '#/components/schemas/Storage'

Storage:
  type: object
  required:
    - type
    - config
    - credentials
  properties:
    type:
      type: string
      description: The type of storage service (e.g., ADLS, S3, GCS)
    config:
      type: object
      description: Configuration settings specific to the storage type
      properties:
        region:
          type: string
          description: The region where the storage is located
        # Add other config properties as needed
    credentials:
      $ref: '#/components/schemas/Credentials'

```

Reply:
@flyrain are you suggesting to standardize all storage-related configuration into their own properties inside `config`? Wouldn't that require a spec change whenever a new configuration option would be added?

Reply:
That's right. We can standardize some of them, like aws region, while still having  a flexible map  to hold abitrary key-value pairs, like this:
```
        extraConfig:
          type: object
          additionalProperties:
            type: string
```

Reply:
Where would we draw the line between which should be standardized and which shouldn't? 
What happens if certain storage configurations become obsolete and get removed from the actual impl? My concern here is that we would have to support those properties forever in the REST spec for older clients.

I'm ok introducing `Storage` as shown below to provide more flexibility with the configuration but I believe I mentioned somewhere else that it's probably better to handle configuration standardization in a separate proposal. @flyrain would you be interested in working on that proposal?

Reply:
Hi @nastra , sorry for the delay. Yes, we could handle it in a separated PR. There is one minor point, that the structure I proposed can reuse the field `type`. For example, this is the current PR:
```
credential
   - type(s3/adsl/gcs)
```
This is the proposal with storage, in which we don't need to specify the storage type.
```
storage
  - type(s3/adsl/gcs)
  - credentials
  - config     <- we can work on this a bit later
```
WDYT?

Reply:
@flyrain I gave this some more thought and I don't think we should be standardizing any of the storage-related configuration in the REST spec for the following reasons:
* those are all Java-specific and other languages might use other property names for the respective FileIO implementation
* I think the issue is rather with documentation, since most storage-related properties aren't well documented. That means we should document those on the website and maybe in the REST spec (similar to what https://github.com/apache/iceberg/pull/10576 attempted). Documenting something in the REST spec is still different than actually defining a schema + properties for that configuration.
 
Due to that I don't think it makes sense to introduce `storage` right now in order to duplicate and provide a subset of the `config` that's already defined in `LoadTableResponse`

Reply:
> those are all Java-specific and other languages might use other property names for the respective FileIO implementation

I didn’t quite get that. Clients in other languages should stick to the names defined in the spec. I can’t imagine the server responding with different config keys, like the S3 region, just because the client is implemented in a different language.

Thanks for sharing the PR to add more comments. I'm OK with **non-stardadrized** **schema-less** for configs as they may change, but I still think it is more organized with one more layer like the following. As I said,  this is a minor point to make the overall schema a bit cleaner, esp. for new storage type adding in the future. 
```
LoadTableResponse:
  storages:
          type: array
          items:
            $ref: '#/components/schemas/Storage'

storage:
      type: object
      discriminator:
        propertyName: type
        mapping:
          adls: '#/components/schemas/adls'
          gcs: '#/components/schemas/gcs'
          s3: '#/components/schemas/s3'
      oneOf:
         ...

s3
  - type(s3/adsl/gcs)
  - credentials
  - config     <- schema-less key-value pairs, add comments here for storage related configs
```

Reply:
> I didn’t quite get that. Clients in other languages should stick to the names defined in the spec. I can’t imagine the server responding with different config keys, like the S3 region, just because the client is implemented in a different language.

I remember that this was the case for pyiceberg a while ago, but looks like this was addressed by https://github.com/apache/iceberg-python/pull/922.

> I'm OK with non-stardadrized schema-less for configs as they may change, but I still think it is more organized with one more layer like the following. As I said, this is a minor point to make the overall schema a bit cleaner, esp. for new storage type adding in the future.

@flyrain do we have a concrete use case for this or is this rather a nice-to-have? I'm thinking that we most likely don't need this right now. Also in the last Iceberg sync we were talking about v2 version of loadTable, so maybe that fits better with a v2 API when we have a concrete use case for introducing this?

Reply:
Sorry a bit late to this, I was in general neutral about the proposal so did not pay too much attention. 

But the AWS client thing seems worth discussion. Ultimately, it sounds like we want the freedom to configure the FileIO in the way we want **per table**, that includes credentials and region and maybe other things in the future.

Given that we already want to formally standardize all the FileIO properties across languages, and also we are using config map to vend the configurations of those FileIO, then why do we still need to have dedicated credentials fields? It does seem a bit redundant to me just for the sake of not using config properties for a selective subset of fields... 

Yufei's proposal of having a dedicated storage field could be a more formalized way to say that these configs are only for FileIO, I am neutral on that.

Reply:
> ... then why do we still need to have dedicated credentials fields

@jackye1995 we want to standardize credentials because of a follow-up proposal that will focus on refreshing vended credentials

Reply:
Is that published? If not, could you give a short overview if possible?

Reply:
> It does seem a bit redundant to me just for the sake of not using config properties for a selective subset of fields...

I think this PR is to apply schema to file IO related fields mainly for credentials, so that, both server impl. and clients can follow without worrying too much about the compatibility, like key names changes.

Reply:
@jackye1995 it's not published yet because I wanted to get this proposal in first and otherwise it would deviate the discussion. 
Here's a quick summary of how refreshing vended credentials would look like:

Refreshing vended credentials would entail adding a new `/v1/{prefix}/namespaces/{namespace}/tables/{table}/credentials` endpoint that returns a `LoadCredentialsResponse`, which carries
```
storage-credentials:
  type: array
  items:
    $ref: '#/components/schemas/Credential'
```
This new endpoint would be called by the mechanism that is offered by the respective storage provider. For example, GCS offers a `OAuth2CredentialsWithRefresh` which then effectively would call this new endpoint to refresh a vended credential.
Achieving the same thing for S3 would look different but effectively comes down to calling this credential endpoint too.

Reply:
Cool, in that case I think this is a great step, it does make a lot of sense to have dedicated credentials refresh API, we were also thinking about proposing it later. We have implemented credentials vending in LakeFormation for a long time now, and last year S3 also offered access grant with similar feature. The user pattern has been proven that having a dedicated API for that makes a lot of sense since refreshing credentials is a much more frequent operation than loading table and can be done very quickly with a dedicated API, and from security perspective it is also a better posture. In that case I think what is proposed here makes sense. In that case, I am more in favor of the current design than what Yufei suggests, because it makes it clear that these fields are related to that new API, and existing config map continues to work for configuring the FileIO, with the additional note that the new credentials field would override the credentials config in the config map.

Reply:
@nastra, thanks for the context. The suggest I provide is a minor structure change, which isn't needed in the new endpoint. I am OK with the current one if that's more align with the new endpoint.